### ISSUE TITLE
`pluto` chooses wrong raw RSA key

### ISSUE THREAD

**Issue by kavinda-forcepoint** [libreswan/libreswan#352]
### Summary

`pluto` chooses to send the wrong RSA key when all of following conditions are satisfied:

* There are 2 raw RSA keys **X1** and **X2** stored inside the NSS (Network Security Services) database.
* Number of bits in **X1** = 3616
* Number of bits in **X2** = 4000
* The secrets file for **X1** is `/etc/ipsec.d/X1.secrets`.
* The secrets file for **X2** is `/etc/ipsec.d/X2.secrets`.
* Connection config **Alpha** uses **X2** for its raw RSA key with `authby=rsasig`.
* **Alpha** uses IKEv1.
* _No_ connection configs use **X1**.

### Configuration

* **Libreswan Version:** 3.29
* **OS:** Red Hat Enterprise Linux 8.1 (64-bit)
* **FIPS Mode:** Enabled
* **Connection Mode:** Tunnel

Here's the connection configuration for both endpoints of the tunnel:

```
conn Alpha
    ikev2=no
    ike=aes256-sha1
    phase2alg=aes256-sha1
    leftid=@libreswan1
    left=<left's IP address>
    leftrsasigkey=<RSA key X1>
    rightid=@libreswan2
    right=<right's IP address>
    rightrsasigkey=<right's RSA key>
    authby=rsasig
    auto=start
```

### Steps to Reproduce

1. Create a new RSA key **X1** with `ipsec newhostkey -n 3616` and save it to `/etc/ipsec.d/X1.secrets`.
2. Create another RSA key **X2** with `ipsec newhostkey -n 4000` and save it to `/etc/ipsec.d/X2.secrets`.
3. Create a connection config file `/etc/ipsec.d/Alpha.conf` using the above `conn Alpha` configuration. Specify **X2** for `leftrsasigkey`.
4. Start `pluto` at both endpoints of the tunnel.
5. If the `left` endpoint acts as a Responder, then you should see log messages similar to the following at the `left` endpoint:

```
"Alpha" #3 MAIN_R2: XXX: different modulus k (452 vs 500) modulus (452 vs 500) caused a mismatch
"Alpha" #3 MAIN_R2: k did NOT match
"Alpha" #3 MAIN_R2: n did NOT match
"Alpha" #3 MAIN_R2: e did match
"Alpha" #3 MAIN_R2: multiple ipsec.secrets entries with distinct secrets match endpoints: first secret used
"Alpha" #3 MAIN_R2: concluding with best_match=002 best=0x556e7f5908e8 (lineno=1)
"Alpha" #3 MAIN_R2: RSA key <RSA keyid for X1> found
"Alpha" #3 MAIN_R2: RSA_sign_hash: Started using NSS
"Alpha" #3 MAIN_R2: NSS: Authentication to NSS successful
"Alpha" #3 MAIN_R2: RSA_sign_hash: Ended using NSS
"Alpha" #3 MAIN_R2: ***emit ISAKMP Signature Payload:
"Alpha" #3 MAIN_R2:    next payload type: ISAKMP_NEXT_NONE (0x0)
"Alpha" #3 MAIN_R2: next payload chain: setting previous 'ISAKMP Identification Payload (IPsec DOI)'.'next payload type' to current ISAKMP Signature Payload (9:ISAKMP_NEXT_SIG)
"Alpha" #3 MAIN_R2: next payload chain: saving location 'ISAKMP Signature Payload'.'next payload type' in 'reply packet'
"Alpha" #3 MAIN_R2: emitting 452 raw bytes of SIG_R into ISAKMP Signature Payload
```

As you can see in the line `RSA key <RSA keyid for X1> found`, `pluto` picks **X1** instead of **X2** as specified in the connection configuration **Alpha** to send to the Initiator.

### Details

The bug happens in the following function:

```
1   struct secret *lsw_find_secret_by_id(struct secret *secrets,
2   				enum PrivateKeyKind kind,
3   				const struct id *my_id,
4   				const struct id *his_id,
5   				bool asym)
6   {
7   	enum {
8   		match_none = 000,
9   
10  		/* bits */
11  		match_default = 001,
12  		match_any = 002,
13  		match_him = 004,
14  		match_me = 010
15  	};
16  	unsigned int best_match = match_none;
17  	struct secret *best = NULL;
18  
19  	<code omitted>	
20  
21  	for (struct secret *s = secrets; s != NULL; s = s->next) {
22  		<code omitted>
23  
24  		if (s->pks.kind == kind) {
25  			unsigned int match = match_none;
26  
27  			if (s->ids == NULL) {
28  				<code omitted>
29  			} else {
30  				/* check if both ends match ids */
31  				struct id_list *i;
32  				int idnum = 0;
33  
34  				for (i = s->ids; i != NULL; i = i->next) {
35  					<code omitted>
36  					<NOTE 1: Since `i->id->kind == ID_NONE`, any_id() returns true>
37  					if (any_id(&i->id)) {
38  						/*
39  						 * match any will automatically
40  						 * match me and him so treat it
41  						 * as its own match type so
42  						 * that specific matches get
43  						 * a higher "match" value and
44  						 * are used in preference to
45  						 * "any" matches.
46  						 */
47  						match |= match_any;
48  					} else {
49  						<code omitted>
50  					}
51  					<code omitted>
52  				}
53  			}
54  			<code omitted>
55  			switch (match) {
56  			<code omitted>
57  			case match_any:	/* a wildcard */
58  			<code omitted>
59  				if (match == best_match) {
60  					/*
61  					 * two good matches are equally good:
62  					 * do they agree?
63  					 */
64  					bool same = FALSE;
65  
66  					switch (kind) {
67  					<code omitted>
68  					case PKK_RSA:
69  						/*
70  						 * Dirty trick: since we have
71  						 * code to compare RSA public
72  						 * keys, but not private keys,
73  						 * we make the assumption that
74  						 * equal public keys mean equal
75  						 * private keys. This ought to
76  						 * work.
77  						 */
78  						<NOTE 2: same_RSA_public_key() returns false for s = X1>
79  						same = same_RSA_public_key(
80  							&s->pks.u.RSA_private_key.pub,
81  							&best->pks.u.RSA_private_key.pub);
82  						break;
83  					<code omitted>
84  					}
85  					if (!same) {
86  						DBG(DBG_CONTROL,
87  						    DBG_log("multiple ipsec.secrets entries with distinct secrets match endpoints: first secret used"));
88  						/*
89  						 * list is backwards:
90  						 * take latest in list
91  						 */
92  						best = s;
93  					}
94  				} else if (match > best_match) {
95  					<code omitted>
96  				} else {
97  					<code omitted>
98  				}
99  			}
100 		}
101 	}
102 }
```

The `secrets` argument to the function is a list of RSA keys **[X2, X1]** in that order.

The `for` loop on line 21 goes through the `secrets` list.

`<NOTE 1>` on line 36 applies to both **X2** and **X1** since they both have `i->id->kind == ID_NONE` for the `ids` stored in them. This means `match |= match_any;` is executed.

The 1st iteration of the `for` loop in line 21:

* Processes **X2** as `s`
* Sets `match` to `match_any` on line 47
* Sets `best` to `s`, i.e. **X2**
* Sets `best_match` to `match_any`

The 2nd iteration of the `for` loop in line 21:

* Processes **X1** as `s`
* Sets `match` to `match_any` on line 47  
* Sets `same` to `false` because `same_RSA_public_key(X1, X2)` returns `false` on line 79 (i.e. `<NOTE 2>`)
* Sets `best` to `s`, i.e. **X1** on line 92

And so, `pluto` ends up using **X1** as the RSA key for connection **Alpha** instead of **X2** as specified.

**Comment by cagney** [comment]
this should be fixed in 4.0 (the changes are in mainline)

left will use the raw private key matching the CKAID of leftrsasigkey

**Comment by letoams** [comment]
thanks. added note to CHANGES file for this.
